### ISSUE TITLE
Lots of calls to _dl_relocate_object

### ISSUE THREAD

**Issue by stevenagy** [dyninst/dyninst#470]
Hello,

I'm using Dyninst to instrument a single basic block, but noticed that the resulting binary incurs some pretty significant overhead on short-running executions. I did some profiling with intel processor trace and noticed a large amount of time being spent on calls to **_dl_relocate_object**. 

My question is - since I'm only interested in instrumenting a basic hook-and-return type of instrumentation (i.e. I insert a callback that executes and returns to the caller), is there any mechanism for me to simplify the amount of work Dyninst seems to be doing in the background? 

Thanks,
-Steve

**Comment by mxz297** [comment]
I am not aware of anything Dyninst does that would cause excessive calls to _dl_relocate_object. Can your provide the calling context of _dl_relocate_object so that I can know what Dyninst operations triggered _dl_relocate_object? Hopefully the profiler you are using can provide you such information.

**Comment by stevenagy** [comment]
My C binary is instrumented in the first basic block of <_init> with a library callback. 
I've attached the perf report output - I'm not exactly sure how to provide you with calling context information. 
[perf.log](https://github.com/dyninst/dyninst/files/2128353/perf.log)

**Comment by mxz297** [comment]
From your data, 76.53% of the total runtime is spent in dl_main. I assume you are using binary rewriting. So, I guess Dyninst is generating strange relocation entries to cause the loader to spend this large amount of time. Can you send me the original binary and rewritten binary?

**Comment by stevenagy** [comment]
Yeah I'm attempting rewriting. I've attached the original and instrumented binaries. 

To run the instrumented one, please make sure you do the following:
`export LD_LIBRARY_PATH=[path to dir where you unzipped the files]`
`export DYNINSTAPI_RT_LIB=[path to dyninst build dir]`

[cjson_dyninst.zip](https://github.com/dyninst/dyninst/files/2129192/cjson_dyninst.zip)

**Comment by mxz297** [comment]
Hi, I was not able to run your rewritten binary because of a hard-coded dependency:
/mnt/bigdata/dynBuildDir/lib/libdyninstAPI_RT.so => not found

Somehow there is another dependency of libdyninstAPI_RT.so, which is

libdyninstAPI_RT.so => /p/paradyn/development/xmeng/dyninstapi/dyninst_development/x86_64-unknown-linux2.4/lib/libdyninstAPI_RT.so (0x00007f8c7b73f000)

So, if you can remove the hard coded dependency, I can try to run and see what is going on

**Comment by stevenagy** [comment]
It might just be easier for me to send you the .cpp's and Makefile. Do you have an email I can forward them to? I'd like to avoid my source code being posted publicly.

**Comment by mxz297** [comment]
Yes, xmeng@cs.wisc.edu

**Comment by stevenagy** [comment]
Sent. Thank you!

**Comment by n738cu** [comment]
@mxz297 : What's the status of this issue?

**Comment by mxz297** [comment]
@n738cu I did not get a chance to work on this.
